### ISSUE TITLE
Leaf value is implicitly deleted even if it is present in the xml config file.

### ISSUE THREAD

**Issue by jabilmplane** [sysrepo/sysrepo#1946]
Hi.

I'm doing some testing for sysrepo/libnetconf2/libyang/Netopeer2. And it seems that I have found several issues on the devel branch.

YANG model is:
```
> get-schema --model test
MODULE
module test {
  namespace "urn:test";
  prefix e;

  container test {
    container cont1 {
      must "leaf1 != 33 and leaf1 != 66" {
        error-message
          "Values 33 and 66 are not valid";
      }
      leaf leaf1 {
        type int8;
        mandatory true;
      }
    }

    leaf leaf2 {
      when "../cont1/leaf1 > 10";
      type int8 {
        range "0..100";
      }
    }

  }
}
```

**Steps to reproduce.**
**1. Install yang model. Set some startup value for leaf1 with sysrepocfg -W**
**2. Create leaf1 and leaf2. test_cfg.xml file is:**
```
<test xmlns="urn:test">
    <cont1>
        <leaf1>11</leaf1>
    </cont1>
    <leaf2>21</leaf2>
</test>
```

**3. Apply configuration. Verify that both leaves have been created**
```
> edit-config --target running --config=/home/test/test_cfg.xml 
OK
> get-config --source running --filter-xpath /test:test 
DATA
<test xmlns="urn:test">
  <cont1>
    <leaf1>11</leaf1>
  </cont1>
  <leaf2>21</leaf2>
</test>
```

**4. Change leaf1 value to 1.
   Leaf 2 value remain the same.
   This configuration is not valid because leaf2 shoud exist only when leaf1 value is more than 10.
   test_cfg.xml file is:**
```
<test xmlns="urn:test">
    <cont1>
        <leaf1>1</leaf1>
    </cont1>
    <leaf2>21</leaf2>
</test>
```

**5. Apply configuration. Leaf2 is implicitly deleted by the server although it is present in xml file.**

`> edit-config --target running --config=/home/test/test_cfg.xml `
**Current behavior**
```
OK
> get-config --source running --filter-xpath /test:test 
DATA
<test xmlns="urn:test">
  <cont1>
    <leaf1>1</leaf1>
  </cont1>
</test>
```
**Expected behavior**
```
ERROR
	type:     application
	tag:      operation-failed
	severity: error
	path:     /test:test/leaf2
	message:  When condition "../cont1/leaf1 > 10" not satisfied.
```

Note. If I call edit-config again, it returns error I expect.

Thanks in advance!

**Comment by michalvasko** [comment]
Hi,
this is actually the correct behavior. Even though you are sending the `leaf2` value in the XML, it already exists in the datastore, so it is ignored for the final validation. When you perform the edit-config the second time, the `leaf2` is created and so you get the error.

Regards,
Michal

**Comment by jabilmplane** [comment]
Yes, my understanding is the same.
But it looks very strange when non valid xml file pases validation when it is applied to the datastore.

**Comment by michalvasko** [comment]
Hi,
well, there is nothing that can be done about this, it is how it functions. The point here is mostly that not an XML file is being applied but rather this file is compared to the current datastore contents, the minimal set of changes is collected, and that is what is actually applied. Refer to the NETCONF [edit-config](https://tools.ietf.org/html/rfc6241#section-7.2) operation for specifics.

Regards,
Michal

**Comment by jabilmplane** [comment]
Thanks, closing it.

**Comment by jabilmplane** [comment]
Hi @michalvasko 

Reopening this ticket.

I have changed leaf2 value to 22 in step 4. So, in the request both values are changed. Although configuration is invalid, edit-config operation returns OK and leaf2 is implicitly deleted.


YANG model is:
```
> get-schema --model test

MODULE
module test {
  namespace "urn:test";
  prefix e;

  container test {
    container cont1 {
      must "leaf1 != 33 and leaf1 != 66" {
        error-message
          "Values 33 and 66 are not valid";
      }
      leaf leaf1 {
        type int8;
        mandatory true;
      }
    }

    leaf leaf2 {
      when "../cont1/leaf1 > 10";
      type int8 {
        range "0..100";
      }
    }

  }
}
```

Steps to reproduce.
1. Install yang model. Set some startup value for leaf1 with sysrepocfg -W
2. Create leaf1 and leaf2. test_cfg.xml file is:
```
<test xmlns="urn:test">
    <cont1>
        <leaf1>11</leaf1>
    </cont1>
    <leaf2>21</leaf2>
</test>
```
3. Apply configuration. Verify that both leaves have been created
```
> edit-config --target running --config=/home/test/test_cfg.xml 
OK
> get-config --source running --filter-xpath /test:test 
DATA
<test xmlns="urn:test">
  <cont1>
    <leaf1>11</leaf1>
  </cont1>
  <leaf2>21</leaf2>
</test>
```
4. Change leaf1 value to 1 and leaf2 value to 22.
This configuration is not valid because leaf2 shoud exist only when leaf1 value is more than 10.
test_cfg.xml file is:
```
<test xmlns="urn:test">
    <cont1>
        <leaf1>1</leaf1>
    </cont1>
    <leaf2>22</leaf2>
</test>
```
5. Apply configuration. Leaf2 is implicitly deleted by the server although new value is present in xml file.

> edit-config --target running --config=/home/test/test_cfg.xml
Current behavior
```
OK
> get-config --source running --filter-xpath /test:test 
DATA
<test xmlns="urn:test">
  <cont1>
    <leaf1>1</leaf1>
  </cont1>
</test>
```
Expected behavior
```
ERROR
	type:     application
	tag:      operation-failed
	severity: error
	path:     /test:test/leaf2
	message:  When condition "../cont1/leaf1 > 10" not satisfied.
```

Thanks in advance!

**Comment by michalvasko** [comment]
Hi,
based on the RFC7950 specification, I think this is the correct behavior. Why do you think it is not?

Regards,
Michal

**Comment by jabilmplane** [comment]
Hi.

Because passed xml file is not valid:
- in Yang there is when statement: when "../cont1/leaf1 > 10";
- xml file is invalid: leaf1 = 1, leaf2 = 22 (new value)

Please, pay attention that before running edit-config, leaf2 has value 21. Then I try to change it to the 22 and also set leaf1 value to 1 to make leaf2 value not valid.

**Comment by michalvasko** [comment]
Hi,
I know the value is invalid in the file, but you are performing an `edit-config`, so all the constraints are specified [here](https://tools.ietf.org/html/rfc7950#section-8.3.2). The last bullet could use some improvement and the first sentence should actually be
> Modification requests for nodes tagged with "when", and the "when"
      condition evaluates to "false" in the current datastore.

Otherwise you could never perform an `edit-config` on any nodes with `when` without the required data always included in the `edit-config` itself.

Regards,
Michal

**Comment by jabilmplane** [comment]
Yes, absolutelly agree that piece of configuration should not be validated separately from the datastore. It is absolutelly clear for me.

But this configuration should be applied to the datastore and validated together with the datastore.

I will try to explain my concern again...

1. In the YANG file there are two leaves: leaf1 and leaf2.
2. leaf2 has "when ../cont1/leaf1 > 10"" statement
3. Before performing edit-config operation, running datastore has:
    leaf1 = 11
    leaf2 = 21
4. In XML config file there are two NEW values:
    leaf1 = 1
    leaf2 = 22 (not valid because of  when statement)

When adit-config is performed, I expect an ERROR because changes should be validated in context of the datastore.
**It is not clear for me why leaf2 is deleted even if it has new value in the config.**

Note that if I run edit-config again, I get an expected error. But I do believe this error should be returned after first edit-config operation.

I could imagine that changes from the config file are processed one-by-one. And if leaf2 is processed first, and then leaf1 is processed, I could imagine following scenario:
    leaf2 value changed from 21 to 22
    leaf1 value changed from 11 to 1. As a result, when expression evaluates to false and leaf2 is deleted.
If my assumption is correct, there is an incosistency in the behavior because final result depends on the changes order.

**Comment by michalvasko** [comment]
Hi,
this is the way I see it.

1. You start with `leaf1` `11` and `leaf2` `21`.
2. You send the `edit-config` with `leaf1` `1` and `leaf2` `22`.
    Edit-config itself is OK, we agreed on that.
3. After `edit-config` is fully (no sub-validations) applied, the whole `running` datastore is validated ([ref](https://tools.ietf.org/html/rfc7950#section-8.3.3)).
4. During validation, it is learned that a `when` of a node became `false`, so the node is removed ([ref](https://tools.ietf.org/html/rfc7950#section-8.2)).

Well, that is all, basically, it does not matter whether `leaf2` has a new value or not, it only matters that its `when` was changed from `true` to `false`.

Also, you are right that if you split these changes into 2 `edit-config`s, the behavior would be different. If you first change `leaf1` to `1`, the validation will cause `leaf2` to be removed and when you then want to change `leaf2` to `22`, you should get `unknown element` error ([ref](https://tools.ietf.org/html/rfc7950#section-8.3.2)).

Please, be more specific with references to the RFC if you still think there is any problem or inconsistency, I just do not see it.

Regards,
Michal
